server/src/storage/uncompressed_data_buffer.rs

```diff
@@ -0,0 +1,423 @@
+    fn get_univariate_id(&self) -> u64 {
+        self.univariate_id
+    }
+
+    /// Return the total size of the [`UncompressedInMemoryDataBuffer`] in bytes.
+    fn get_memory_size(&self) -> usize {
+        UncompressedInMemoryDataBuffer::get_memory_size()
+    }
+
+    /// Spill the in-memory [`UncompressedInMemoryDataBuffer`] to an Apache Parquet file and return
+    /// the [`UncompressedOnDiskDataBuffer`] when finished.
+    fn spill_to_apache_parquet(
+        &mut self,
+        local_data_folder: &Path,
+        uncompressed_schema: &UncompressedSchema,
+    ) -> Result<UncompressedOnDiskDataBuffer, IOError> {
+        // Since the schema is constant and the columns are always the same length, creating the
+        // RecordBatch should never fail and unwrap is therefore safe to use.
+        let batch = self.get_record_batch(uncompressed_schema).unwrap();
+        UncompressedOnDiskDataBuffer::new(self.univariate_id, local_data_folder, batch)
+    }
+}
+
+/// A read only uncompressed buffer that has been spilled to disk as an Apache Parquet file due to
+/// memory constraints.
+pub struct UncompressedOnDiskDataBuffer {
+    /// Id that uniquely identifies the time series the buffer stores data points from.
+    univariate_id: u64,
+    /// Path to the Apache Parquet file containing the uncompressed data in the
+    /// [`UncompressedOnDiskDataBuffer`].
+    file_path: PathBuf,
+}
+
+impl UncompressedOnDiskDataBuffer {
+    /// Spill the in-memory `data_points` from the time series with `univariate_id` to an Apache
+    /// Parquet file in `local_data_folder`. If the Apache Parquet file is written successfully,
+    /// return an [`UncompressedOnDiskDataBuffer`], otherwise return [`IOError`].
+    pub(super) fn new(
+        univariate_id: u64,
+        local_data_folder: &Path,
+        data_points: RecordBatch,
+    ) -> Result<Self, IOError> {
+        let local_file_path = local_data_folder
+            .join("uncompressed")
+            .join(univariate_id.to_string());
+
+        // Create the folder structure if it does not already exist.
+        fs::create_dir_all(local_file_path.as_path())?;
+
+        // Create a path that uses the first timestamp as the filename.
+        let timestamps = get_array!(data_points, 0, TimestampArray);
+        let file_name = format!("{}.parquet", timestamps.value(0));
+        let file_path = local_file_path.join(file_name);
+
+        StorageEngine::write_batch_to_apache_parquet_file(data_points, file_path.as_path())?;
+
+        Ok(Self {
+            univariate_id,
+            file_path,
+        })
+    }
+}
+
+impl fmt::Debug for UncompressedOnDiskDataBuffer {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(
+            f,
+            "UncompressedOnDiskDataBuffer({}, {})",
+            self.univariate_id,
+            self.file_path.display()
+        )
+    }
+}
+
+// TODO: Delete the spilled Apache Parquet file after it has been read?
```

Comment:
when does the file get deleted now?

Reply:
Currently they don't get deleted as far as I am aware as the [`UncompressedDataManager`](https://github.com/ModelarData/ModelarDB-RS/blob/dev/optimize-storage-engine/server/src/storage/uncompressed_data_manager.rs) interacts with [`UncompressedInMemoryDataBuffer`](https://github.com/ModelarData/ModelarDB-RS/blob/dev/optimize-storage-engine/server/src/storage/uncompressed_data_buffer.rs#L72) and [`UncompressedOnDiskDataBuffer`](https://github.com/ModelarData/ModelarDB-RS/blob/dev/optimize-storage-engine/server/src/storage/uncompressed_data_buffer.rs#L184) through [`UncompressedDataBuffer`](https://github.com/ModelarData/ModelarDB-RS/blob/dev/optimize-storage-engine/server/src/storage/uncompressed_data_buffer.rs#L44) and [`UncompressedOnDiskDataBuffer.get_record_batch()`](https://github.com/ModelarData/ModelarDB-RS/blob/dev/optimize-storage-engine/server/src/storage/uncompressed_data_buffer.rs#L236) does not delete the file after returning its contents as a [`RecordBatch`](https://docs.rs/arrow/latest/arrow/record_batch/struct.RecordBatch.html).

Reply:
I added the TODO instead of adding code to delete the file as it seemed outside the scope of this PR since it requires creating a component to properly recover from abnormal termination to ensure the files are always deleted, e.g., by reading the files into memory if any exists when the system starts.